File: apps/api/plane/app/serializers/agent.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

"""
Serializers for the workspace-agent actor type - category 9, feature 7
(docs/feature-specs/09-ai-features.md "7. Type d'acteur agent de premiere
classe" in plane-selfhost).
"""

from rest_framework import serializers

from plane.db.models import AgentProfile, APIToken

from .base import BaseSerializer
from .user import UserLiteSerializer


class AgentProfileSerializer(BaseSerializer):
    user_id = serializers.PrimaryKeyRelatedField(source="user", read_only=True)
    display_name = serializers.CharField(source="user.display_name", read_only=True)
    email = serializers.CharField(source="user.email", read_only=True)
    avatar_url = serializers.CharField(source="user.avatar_url", read_only=True)
    created_by = UserLiteSerializer(read_only=True)
    # Annotated on the queryset (see AgentProfileViewSet.get_queryset) -
    # exigence "nombre de projets ou il est membre" (spec's own "Considerations
    # API/UX" section, GET /workspaces/{slug}/agents/).
    project_count = serializers.IntegerField(read_only=True, default=0)

    class Meta:
        model = AgentProfile
        fields = [
            "id",
            "user_id",
            "display_name",
            "email",
            "avatar_url",
            "workspace",
            "agent_type",
            "status",
            "description",
            "created_by",
            "last_seen_at",
            "created_at",
            "updated_at",
            "project_count",
        ]
        read_only_fields = [
            "id",
            "user_id",
            "display_name",
            "email",
            "avatar_url",
            "workspace",
            "created_by",
            "last_seen_at",
            "created_at",
            "updated_at",
            "project_count",
        ]


class AgentAPITokenSerializer(BaseSerializer):
    """Same shape as the personal `APITokenSerializer` - includes the raw
    `token` value, so this is ONLY ever used for the issuance response
    (`POST .../tokens/`), matching the existing personal-token UX
    (`plane.app.views.api.ApiTokenEndpoint`). Listing existing tokens must
    use `AgentAPITokenReadSerializer` below instead - `read_only_fields`
    only gates writes, it does not hide a field from serialized output."""

    class Meta:
        model = APIToken
        fields = "__all__"
        read_only_fields = [
            "id",
            "token",
            "expired_at",
            "created_at",
            "updated_at",
            "workspace",
            "user",
            "user_type",
            "agent",
            "is_active",
            "last_used",
            "rate_limit_tier",
            "rate_limit_overridden_by",
            "rate_limit_overridden_at",
            "rate_limit_override_reason",
        ]


class AgentAPITokenReadSerializer(BaseSerializer):
    """Same field set as `AgentAPITokenSerializer` minus the raw `token`
    value - used for `GET .../tokens/` so a past token is never
    re-exposed after issuance, mirroring `plane.app.serializers.api.
    APITokenReadSerializer`."""

    class Meta:
        model = APIToken
        exclude = ("token",)

File: packages/types/src/agent.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

// Types for the first-class workspace-agent actor type - category 9,
// feature 7 (docs/feature-specs/09-ai-features.md "7. Type d'acteur agent
// de premiere classe" in plane-selfhost), and the backend half of this
// feature: apps/api/plane/app/views/agent.py,
// apps/api/plane/app/serializers/agent.py, apps/api/plane/db/models/agent.py.
// `AgentProfile` sits on top of a bot `User` row (`is_bot=True`,
// `bot_type="WORKSPACE_AGENT"` - see `WORKSPACE_AGENT_BOT_TYPE` in
// @plane/utils and `IUserLite.bot_type` in ./users), which is why
// `IAgentProfile` itself carries no `is_bot`/`bot_type` field of its own.

import type { IUserLite } from "./users";

export type TAgentType = "claude_code" | "cursor" | "generic";

export type TAgentStatus = "ACTIVE" | "DISABLED";

/** Mirrors `AgentProfileSerializer` exactly (apps/api/plane/app/serializers/
 * agent.py) - every field here is read-only on the wire except through the
 * dedicated `TAgentCreatePayload`/`TAgentUpdatePayload` request shapes
 * below; `AgentProfileSerializer.Meta.read_only_fields` marks everything
 * except `agent_type`/`description`/`status` as server-derived. */
export type IAgentProfile = {
  id: string;
  user_id: string;
  display_name: string;
  email: string;
  avatar_url: string;
  workspace: string;
  agent_type: TAgentType;
  status: TAgentStatus;
  description: string;
  created_by: IUserLite | null;
  last_seen_at: string | null;
  created_at: string;
  updated_at: string;
  /** Annotated on the backend queryset - active `ProjectMember` count. */
  project_count: number;
};

/** Body of `POST /api/workspaces/{slug}/agents/`
 * (`AgentProfileViewSet.create`) - note there is deliberately no `role`
 * field: an agent is always created as workspace Member, never Admin, and
 * the backend never reads a `role` key off the request body at all. */
export type TAgentCreatePayload = {
  display_name: string;
  agent_type: TAgentType;
  description?: string;
};

/** Body of `PATCH /api/workspaces/{slug}/agents/{id}/`
 * (`AgentProfileViewSet.partial_update`) - every key is optional and
 * independently applied; setting `status: "DISABLED"` immediately revokes
 * every active token for this agent server-side. */
export type TAgentUpdatePayload = Partial<{
  display_name: string;
  description: string;
  agent_type: TAgentType;
  status: TAgentStatus;
}>;

/** Shape returned by both agent-token endpoints
 * (`AgentTokenListCreateEndpoint`, apps/api/plane/app/views/agent.py) -
 * `token` (the raw secret) is present ONLY on the `POST` (issuance)
 * response (`AgentAPITokenSerializer`); `GET` (list) uses
 * `AgentAPITokenReadSerializer`, which excludes it - same "shown once,
 * never again" convention as the pre-existing personal `IApiToken`
 * (./api_token.ts). */
export type TAgentAPIToken = {
  id: string;
  label: string;
  description: string;
  is_active: boolean;
  last_used: string | null;
  created_at: string;
  updated_at: string;
  expired_at: string | null;
  agent: string;
  workspace: string;
  user: string;
  user_type: number;
  token?: string;
};

/** Body of `POST /api/workspaces/{slug}/agents/{id}/tokens/`
 * (`AgentTokenListCreateEndpoint.post`) - both fields optional, the
 * backend defaults `label` to `"<agent display name> token"` when omitted. */
export type TAgentTokenCreatePayload = Partial<{
  label: string;
  description: string;
}>;
